src/twisted/web/http_headers.py

```diff
@@ -71,8 +72,8 @@ class Headers:
         b"x-xss-protection": b"X-XSS-Protection",
     }
 
-    def __init__(self, rawHeaders=None):
-        self._rawHeaders = {}
+    def __init__(self, rawHeaders: Optional[Dict[String, List[String]]] = None):
```

Comment:
Do we really want `Dict` here, or is this more of a `Mapping`?  Similarly; `List` or `Sequence`?  (We should be careful as we add types not to get too overly specific and make everybody's unit tests need `type: ignore` all over the place; use the least specific type that works for inputs)

Reply:
The docstring for `_rawHeaders` is:

```
   @ivar _rawHeaders: A L{dict} mapping header names as L{bytes} to L{list}s of
        header values as L{bytes}.
```

and the initial variable annotation is:

```
       self._rawHeaders = {}  # type: Dict[bytes, List[bytes]]
```


Is it better to stick with this, or to change the rawHeaders argument to:

```
    def __init__(
        self,
        rawHeaders: Optional[Mapping[Union[bytes, str], Sequence[Union[bytes, str]]]] = None,
    ):
```

Reply:
We should use `Mapping` and `Sequence` here unless the implementation actually expects to be given those specific type (which would be unfortunate).

The type of `_rawHeaders` is internal, so it can (and should be, in this case), more specific than the inputs to `__init__`, so I would not argue that it dictates the answer here.

FWIW I think that Twisted's docstrings a very prone to saying `list` where `Iterable` or `Sequence` is more appropriate, so we should be cleaning that up as we add typing, not preserving it.  In docstrings, the mistake produces no errors. With type checkers, we start to need the accuracy in a stricter sense.

Reply:
Actually, I think `Iterable`, not `Sequence`.  Which necessarily means same in `setRawHeaders`.  Which then also means `_encodeValues` and friends.

Reply:
Actually, I think `Iterable`, not `Sequence`.  Note this necessarily means the same in `setRawHeaders`.  Which then also means `_encodeValues` and friends.

Also, the type check for values in `setRawHeaders` is wrong for this reason… it shouldn't have to be a list; we should be checking that it's an iterable.  That might mean `hasattr(values, "__iter__")` instead.

Reply:
FWIW I'm very prone to using tuples instead of lists, because I don't like mutable types when you don't need mutability, and a class like this should not take in and modify a mutable object without being very clear that it's doing so.  (Which is what using `List` implies.)

Reply:
OK, I changed to `Iterable`.

Reply:
Actually I had to change this back to `Sequence`, because some of the test cases which passed in a `dict` were not raising a `TypeError`